src/dune_pkg/rev_store.ml

```diff
@@ -301,55 +339,188 @@ module File = struct
         match Re.Group.get m 1 with
         | "blob" ->
           Some
-            { hash = Re.Group.get m 2
-            ; size = Int.of_string_exn @@ Re.Group.get m 3
-            ; path = Path.Local.of_string @@ Re.Group.get m 4
-            }
+            (Blob
+               { hash = Re.Group.get m 2
+               ; size = Int.of_string_exn @@ Re.Group.get m 3
+               ; path = Path.Local.of_string @@ Re.Group.get m 4
+               })
+        | "commit" ->
+          Some
+            (Commit
+               { rev = Rev (Re.Group.get m 2)
+               ; path = Path.Local.of_string @@ Re.Group.get m 4
+               })
         | _ -> None)
   ;;
 
-  let path t = t.path
+  let path = function
+    | Blob { path; _ } | Commit { path; _ } -> path
+  ;;
 end
 
+(* a submodule in [.gitmodules] can also have a [branch] but given we only need
+   to resolve the commit
+   object, we don't have to care about the tracking branch *)
+type submodule =
+  { path : Path.Local.t
+  ; source : string
+  }
+
+let parse_submodules =
+  let path_re = Re.(compile @@ seq [ str "path = "; group (rep1 any) ]) in
+  let url_re = Re.(compile @@ seq [ str "url = "; group (rep1 any) ]) in
+  fun gitmodules ->
+    let lines = String.split ~on:'\n' gitmodules in
+    let parse_section lines =
+      let section, rest =
+        List.split_while lines ~f:(fun line -> not @@ String.is_prefix ~prefix:"[" line)
+      in
+      let locate_by_re re line =
+        Re.exec_opt re line |> Option.map ~f:(fun m -> Re.Group.get m 1)
+      in
+      let path = List.find_map section ~f:(locate_by_re path_re) in
+      let url = List.find_map section ~f:(locate_by_re url_re) in
+      match path, url with
+      | Some path, Some source ->
+        let path = Path.Local.of_string path in
+        { path; source }, rest
+      | _, _ -> failwith "Submodule definition missing path & url"
```

Comment:
Should be either Code_error or User_error

Reply:
Indeed, looks like I overlooked this when submitting. However, I wonder about using `Loc.t` for non-dune files: would that be a good idea?

Reply:
If the error corresponds to something that comes from a source file, that would be a good idea. I wouldn't use `Loc.t` for things like internal config files of git for example though.